pip/pip-376.md

```diff
@@ -0,0 +1,138 @@
+
+https://github.com/apache/pulsar/pull/21231 adds two asynchronous overrides that are much more friendly to users:
+
+```java
+CompletableFuture<Optional<TopicPolicies>> getTopicPoliciesAsync(@Nonnull TopicName topicName, boolean isGlobal);
+CompletableFuture<Optional<TopicPolicies>> getTopicPoliciesAsync(@Nonnull TopicName topicName);
+```
+
+Now we have **5** asynchronous get methods. What's worse, unlike `getTopicPolicies`, `getTopicPoliciesAsync(topic)` is not equivalent to `getTopicPoliciesAsync(topic, false)`, instead,
+- `getTopicPoliciesAsync(topic)` will try getting local policies first, if absent, then try getting global policies
+- `getTopicPoliciesAsync(topic, true)` will try getting global policies
+- `getTopicPoliciesAsync(topic, false)` will try getting local policies
+
+It should be noted that the topic policies support global policies across clusters since [#12517](https://github.com/apache/pulsar/pull/12517). So there are local policies and global policies.
+
+Currently,
+- `getTopicPoliciesAsync(TopicName)` is used in `BrokerService#getTopicPoliciesBypassSystemTopic`, which is called when initializing the topic policies of `PersistentTopic` objects. So it uses the "local-first" semantics in case the global policies or local policies is deleted.
+- `getTopicPoliciesAsyncWithRetry` is used in `AdminResource#getTopicPoliciesAsyncWithRetry`, which is called for all topic policies admin APIs. Since these admin APIs all have a `isGlobal` field to indicate whether to get the global policies, it uses the "local only" or "global only" semantics.
+- Other methods are never called directly other than tests.
+
+Actually there is the 6th method `getTopicPoliciesIfExists`, which just tries to get the local topic policies from the cache. This method is the most clear and simple in all these stuffs.
+
+```java
+    TopicPolicies getTopicPoliciesIfExists(TopicName topicName);
+```
+
+# Motivation
+
+Make `TopicPoliciesService` pluggable so that users can customize topic policies service via another backend metadata store.
+
+# Goals
+
+## In Scope
+
+Redesign a clear and simple `TopicPoliciesService` interface for users to customize.
+
+# High Level Design
+
+Add a `topicPoliciesServiceClassName` config to specify the topic policies service class name. If the class name is not the default `SystemTopicBasedTopicPoliciesService`, `systemTopicEnabled` will not be required unless the implementation requires it.
+
+# Detailed Design
+
+## Design & Implementation Details
+
+1. Add a unified method to get topic policies.
+
+```java
+    enum GetType {
+        LOCAL_FIRST, // try getting the local topic policies, if not present, then get the global policies
+        GLOBAL_ONLY, // only get the global policies
+        LOCAL_ONLY,  // only get the local policies
+    }
+    CompletableFuture<Optional<TopicPolicies>> getTopicPoliciesAsync(TopicName topicName, GetType type);
+```
```

Comment:
Do we really need to expose LOCAL_FIRST? The priority of the policies defined in Pulsar

local_topic > global_topic > namespace > dymamic_config > broker.conf

Do we want users to customize the priority? If no, we can only let users to implement the method to get local policies and global policies.

Reply:
The issue is actually related to where this API could be called. See the documents before:

> - `getTopicPoliciesAsync(TopicName)` is used in `BrokerService#getTopicPoliciesBypassSystemTopic`, which is called when initializing the topic policies of `PersistentTopic` objects. So it uses the "local-first" semantics in case the global policies or local policies is deleted.
> - `getTopicPoliciesAsyncWithRetry` is used in `AdminResource#getTopicPoliciesAsyncWithRetry`, which is called for all topic policies admin APIs. Since these admin APIs all have a `isGlobal` field to indicate whether to get the global policies, it uses the "local only" or "global only" semantics.

In `BrokerService` needs to read the policies according to the priority: `local > global`. However, for admin APIs, the current behavior is to read either local or global policies.

> Do we want users to customize the priority? 

The answer is no. But users need to support these three types. For example, if the customized topic policies do not support global policies, this method should return a failed future when the type is `GLOBAL_ONLY`.